Feature/reset password

WinnMoo opened the pull request:
# Description

Added all the functions needed to CRUD ResetIDs needed for password reset/update. Sorry for the big chunk of code, but if I didn't do a PR now it'd be much bigger later.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

I haven't been able to perform unit tests or integration tests because the project can't be built. Some user stuff is missing at the moment.

I've only been able to "test" by hand tracing the code and following the different scenarios.


# Notes

This is the first time I've had to do a pull request, so please bear with me. 😭

WinnMoo committed:
Created controller and unit testing files

WinnMoo committed:
Created necessary files and added some functions

WinnMoo committed:
Updated EmailService to focus on sending emails, moved functions from EmailService to ResetService

WinnMoo committed:
Added the model for reset token, created a repository class for reset token

WinnMoo committed:
Added functions to ResetRepository class

No logic yet

WinnMoo committed:
Updated ResetController

WinnMoo committed:
Fleshed out the data access layer for reset tokens

WinnMoo committed:
Completed reset functions

TODO:
* Start on URI code
* Start on UI code

WinnMoo committed:
Added 2 more needed functions to ResetService

TODO:
* Start on URI code
* Start on UI code
* Start Unit Tests for reset tokens

WinnMoo committed:
Refactored emailService, add 2 more unit tests, started work on resetController

WinnMoo committed:
Refactored ResetService, Added code for UnitTests, Added logic for URIs

WinnMoo committed:
Refactored the way the reset functions work

Vong told me I was doing it wrong :neutral_face:

WinnMoo committed:
Broke up the email functions, refactored unittest for reset password

WinnMoo committed:
Added code for updating password, getting security questions, checking security answers

WinnMoo committed:
Added more logic to accomplish business rules

Added more columns for the reset table to keep track of: how many attempts have been made with that resetID, if that resetID is locked, how long that resetID is locked for

Changed some of the user classes to include checking the DB for an existing user

Added some logic to the controller for handling business rules

WinnMoo committed:
Updated all function names to follow naming convention, deleted trivial unit tests

WinnMoo committed:
Reverted targeted framework back to 4.6.2

ChrisMeyer7088 commented on Backend/DataAccessLayer/Models/Reset.cs:
```diff
@@ -0,0 +1,33 @@
+namespace DataAccessLayer.Models
+{
+    public class Reset
+    {
```
Please change model to a more descriptive name, Reset -> PasswordReset

ChrisMeyer7088 commented on Backend/DataAccessLayer/Models/Reset.cs:
```diff
@@ -0,0 +1,33 @@
+    {
+        [Required, Key] 
+        public string resetID { get; set; }
+
```
ResetID will be a GUID

julianpoy replied:
Additionally, resetId even as a GUID is _not crytographically secure._ You'll need an additional field "token" in the database to store a cryptographically secure string, verifying return users.

WinnMoo replied:
I talked to Vong about using tokens for resetting the password. He said something along the lines of tokens not being necessary in this case because tokens are typically used when information stored in it needs to be passed around publicly. Since we have a table for resetting passwords, and all the information about the reset IDs is being stored there just further shows that the token isn't needed for resetting passwords.

I originally was using GUID to create the resetID, but Vong mentioned that it did have to be cryptographically secure so I'm using RNGCryptoServiceProvider to generate the resetIDs.

julianpoy replied:
I think you misunderstood - I'm not talking about a JWT. Simply about a token (a random string of characters).

There needs to be a separation of concerns here. We need two identifiers for each entry in the table. The first is the database ID that will be referenced by other tables. These are GUIDs across the entire application (see https://github.com/CECS-491A/SSO/blob/b4e0fd5c9500ddaae5afc93796b35c1390136787/Backend/DataAccessLayer/Models/User.cs#L13). The second is the token (again, random string) that will be sent to the user in the email (in link form, likely as a query parameter at the end of the link). This token _must_ be hard to guess. GUIDs are not very secure - they are fairly easily guessed.

The way you have it now, some other random person can guess the GUID and gain access to your account.

ChrisMeyer7088 commented on Backend/DataAccessLayer/Repositories/IResetRepository.cs:
```diff
@@ -0,0 +1,27 @@
+        bool CheckLockOut(string resetID);
+        int GetAttemptsPerResetID(string resetID);
+        int GetResetIDCountPerEmail(string email);
+    }
```
Having a general method to return the PasswordReset Object by passing passwordReset ID would save the need for a large amount of the functions as you could retrieve associated variables from the object instead of specific variables.

Add method: PasswordReset getPasswordReset(DatabaseContext _db, GUID passwordResetID)

Only methods really needed are basic crud operations:
PasswordReset createPasswordReset((DatabaseContext _db, PasswordReset newPasswordReset)
PasswordReset getPasswordReset((DatabaseContext _db, GUID passwordResetID)
PasswordReset  deletePasswordReset((DatabaseContext _db, GUID passwordResetID)
PasswordReset updatePasswordReset((DatabaseContext _db, GUID passwordResetID)

ChrisMeyer7088 commented on Backend/DataAccessLayer/Repositories/IResetRepository.cs:
```diff
@@ -0,0 +1,27 @@
+{
+    interface IResetRepository
+    {
+        void AddResetID(string email, string resetID, DateTime expirationTime);
```
method name AddResetID -> createPasswordReset
string resetID -> GUID passwordResetID
There is no email attribute associated without passwordReset object, do you mean GUID userID

WinnMoo replied:
I used email because when the user clicks the link to start the password reset process, they have to supply an email to send the password reset link to. 

The resetID is then created but before being added to the database, the GUID of the user associated with that email is determined.

julianpoy replied:
That determination should be made at a higher layer. AddResetId should take a user GUID, not an email address. As you have it now, you're duplicating functionality from the User management repo.

ChrisMeyer7088 commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+using System.Threading.Tasks;
+using DataAccessLayer.Models;
+using DataAccessLayer.Database;
+
```
Update ResetPasswordRepository to reflect new interface methods

ChrisMeyer7088 commented on Backend/ServiceLayer/Services/ResetService.cs:
```diff
@@ -0,0 +1,229 @@
+            return resetID;
+        }
+
+        public string CreateResetURL(string resetID)
```
All resetID's will be GUID

julianpoy replied:
Nah, all reset URLs will be the "token"

ChrisMeyer7088 commented on Backend/ServiceLayer/Services/ResetService.cs:
```diff
@@ -0,0 +1,229 @@
+        }
+
+        //Function to create token in the system
+        public string CreateResetID()
```
Need to change this to create a new GUID

julianpoy replied:
Nah, this is where he's creating his "token" - See my comments above. GUID (ID) for the object and the "token" need to be two separate things. The ID is used to identify the object within the database, while the token is a cryptographically secure string.

julianpoy replied:
Please take a look at our session service. This type of logic has already been created. If it's not already reusable from there, make it reusable so we only have token generation logic in one place. As of now, this is a duplicate.

ChrisMeyer7088 commented on Backend/ServiceLayer/Services/ResetService.cs:
```diff
@@ -0,0 +1,229 @@
+        public ResetService()
+        {
+            _resetRepo = new ResetRepository(new DatabaseContext());
+        }
```
DatabaseContext is created in manager layer so that changes may be rolled back after several services are called

julianpoy replied:
This applies to a comment I made above

ChrisMeyer7088 commented on Backend/ServiceLayer/Services/ResetService.cs:
```diff
@@ -0,0 +1,229 @@
+
+        //Function to add the token to the database
+        public void AddResetIDToDB(string userName, string resetID)
+        {
```
Service classes should reflect changes in the repository classes

PasswordReset createPasswordReset(GUID userID, DatabaseContext _db) {
    PasswordReset newPasswordReset = {...}
    _passwordResetRepo.createPasswordReset(newPasswordReset, _db);
}

Need services classes for all CRUD operations and useful logic operaitons such as 
boolean isPasswordResetExpired(PasswordReset passwordReset) //checks if given reset has already passed expiration. You can remove redundant methods that the new CRUD methods will cover such as getSecurityQuestions and lockResetID.

ChrisMeyer7088 commented on Backend/UnitTesting/ResetPasswordUT.cs:
```diff
@@ -0,0 +1,40 @@
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System;
+using System.Collections.Generic;
```
Unit Test should test all CRUD operations of PasswordResetService and EmailResetService

julianpoy replied:
Yeah, these tests are very skimpy

ChrisMeyer7088 commented on Backend/WebAPI/Controllers/ResetController.cs:
```diff
@@ -0,0 +1,123 @@
+        [ActionName("SendEmail")]
+        public void SendResetEmail(string email)
+        {
+            ResetService rs = new ResetService();
```
Services should not be called by the API routes, order of operations should go API Routes call the appropriate method in the Manager Layer, Manager Layer is where the Database Context is initalized so that we are able to roll back any changes if anything goes wrong, Manager Layer is in charge of calling all necessary services, services call the repo which make the database changes then return a value. This continues until all necessary services are called by the managerLayer, in which if everything worked out the manager layer will be in charge of db.saveChanges().

ChrisMeyer7088 commented on Backend/WebAPI/Controllers/ResetController.cs:
```diff
@@ -0,0 +1,123 @@
+            ResetService rs = new ResetService();
+            UserManagementManager umm = new UserManagementManager();
+            //Check to see if the email provided is associated with an account
+            if (umm.ExistingUser(email))
```
The route should mainly focus on formatting the data recieved from the body of the request. Most logic should be carried out in the managerlayer.

ChrisMeyer7088 reviewed (dismissed):
Good effort in work! These changes are to be expected as we get use to a standardized way of doing things.

julianpoy commented on Backend/DataAccessLayer/Models/Reset.cs:
```diff
@@ -0,0 +1,33 @@
+
+        [Required]
+        //Variable to keep track of how many attempts were made with this reset ID
+        public int resetCount { get; set; }
```
Not sure I understand the purpose of this count

julianpoy commented on Backend/DataAccessLayer/Models/Reset.cs:
```diff
@@ -0,0 +1,33 @@
+
+        [Required]
+        //Variable to keep track of how many attempts were made with this reset ID
+        public int resetCount { get; set; }
```
Not sure I understand the purpose of this count

WinnMoo replied:
Each resetID will allows up to 3 attempts when the user answers the security questions. If the user isn't able to correctly answer the security questions in 3 tries, then the resetID is marked locked. 

If the resetID is locked, the user is allowed to create a new password reset link for a total of 3 times within 24 hours.

julianpoy replied:
I was under the impression that the system would ask the security questions up front (as I've seen in numerous applications).

Doing it this way means we have to store these attempt counts, introducing further complication into the application. A simpler suggestion would be to have the security questions answered first.

julianpoy commented on Backend/DataAccessLayer/Models/Reset.cs:
```diff
@@ -0,0 +1,33 @@
+
+        [Required]
+        //Variable to keep track of user being locked out of resetting password
+        public bool lockedOut { get; set; }
```
Not sure I understand the purpose of this boolean

julianpoy commented on Backend/DataAccessLayer/Models/Reset.cs:
```diff
@@ -0,0 +1,33 @@
+        public bool lockedOut { get; set; }
+
+        //Variable to keep track of how long the user is locked out for
+        public DateTime lockoutTime { get; set; }
```
Not sure I understand the purpose of this time...

WinnMoo replied:
If the user has created a password reset link 3 times and has failed to answer the security questions, then the user is unable to create a new password reset link for 24 hours. 24 hours is added to the current time and then added to the table.

The system is able to see if the user isn't able to create new passwords by getting the most recent resetID and seeing when the lockoutTime is on that ID.

julianpoy replied:
See my comment above

julianpoy commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+        private DatabaseContext ResetContext;
+        public ResetRepository(DatabaseContext _db)
+        {
+            this.ResetContext = _db;
```
I may not want to instantiate the whole class with a database context. The existing pattern is to have each individual function accept a database context as the first argument.

julianpoy commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+        }
+
+        //Function to add the token to the DB
+        public void AddResetID(string userName, string resetIDToAdd, DateTime expirationTime)
```
This method should be called "createResetID"

Additionally, resetRepository shouldn't be dependent on users like this. Instead of the argument "userName", this method should take the argument "userId"

Lastly, resetId is generated by the _database_, not by the user creating the password reset. This shouldn't be an argument. Same goes for expirationTime.

julianpoy commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+            Guid userID = userToCheckForResetToken.Id;
+            var reset = ResetContext.ResetIDs.Find(userID);
+            return reset.resetID;
+        }
```
The getter methods above all query the database. Instead, centralize that database call into a single getter method, and have several helper methods that all call that central getter method, then return the data. IE:

public Reset GetResetId(Guid resetId) { queries db and returns Reset object }

julianpoy commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+        }
+
+        //Function to see if the token exists in the DB, given the token
+        public bool ExistingResetIDGivenResetID(string resetID)
```
This method is replaced by my comment above. Remove this method

julianpoy commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+        }
+
+        //Function to see if the token exists in the DB, given the username
+        public bool ExistingResetIDGivenEmail(string userName)
```
This method is pointless

julianpoy commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+        }
+
+        //Function to update the password in the DB
+        public bool UpdatePassword(Guid userIDToChangePassword, string newPasswordHash)
```
Remove this method - it is/should be part of the User repository.

julianpoy commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+        }
+
+        //Function to get security questions from the DB
+        public List<string> GetSecurityQuestions(Guid userIDToGetQuestionsFrom)
```
This method should be part of the User repository, not part of the reset repository

julianpoy commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+        }
+
+        //Function to check security answers against the DB
+        public bool CheckSecurityAnswers(Guid userIDToGetQuestionsFrom, List<string> userSubmittedSecurityAnswers)
```
Same as above

julianpoy commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+        }
+
+        //Function to toggle lockout
+        public void LockOut(string resetID)
```
What does lockout do? I think you're confusing functionality. ResetId should be a token that is generated after the user presses "reset" and enters their email. Login attemps and lockouts should be tracked elsewhere.

julianpoy commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+        }
+
+        //Function to check if locked out, returns true if still locked out, false if not locked out
+        public bool CheckLockOut(string resetID)
```
Same as above

julianpoy commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+        }
+
+        //Function to get the amount of times the resetID has been used to attempt a password reset
+        public int GetAttemptsPerResetID(string resetID)
```
Same as above

julianpoy commented on Backend/DataAccessLayer/Repositories/ResetRepository.cs:
```diff
@@ -0,0 +1,232 @@
+            return 3;
+        }
+
+        public int GetResetIDCountPerEmail(string email)
```
Same as above

julianpoy commented on Backend/DataAccessLayer/Repositories/UserRepository.cs:
```diff
@@ -23,5 +23,7 @@ public bool ValidatePassword(User user, string passwordSubmittedHash)
             }
         }
+
+
```
Please revert these changes if you're not going to make changes here.

julianpoy commented on Backend/ManagerLayer/UserManagement/UserManagementManager.cs:
```diff
@@ -13,7 +13,10 @@ namespace ManagerLayer.UserManagement
-        private IUserService _userService;
+        //private IUserService _userService;
+        //I changed _userService to type UserService because IUserService is only CRUD operations
+        //I also implemented existingUser in UserService to make use of the method available in UserManagementRepository -Winn
```
These comments should not be inline. Although okay if you'd like to make them on your PR, please do not leave them in the source.

Lastly, please leave this as it was. You should be able to add your method to the interface.

julianpoy commented on Backend/ManagerLayer/UserManagement/UserManagementManager.cs:
```diff
@@ -141,5 +144,13 @@ public int UpdateUser(User user)
             }
         }
+
+        public bool ExistingUser(string email)
```
Check the user repository - does this/a similar method already exist? I'm not sure, but probably best to check on that.

julianpoy commented on Backend/ServiceLayer/ServiceLayer.csproj:
```diff
@@ -49,11 +60,14 @@
     <Compile Include="Services\ISessionService.cs" />
     <Compile Include="Services\IUserService.cs" />
     <Compile Include="Services\PasswordService.cs" />
+    <Compile Include="Services\ResetService.cs" />
```
Reset service needs an interface (IResetService)

julianpoy commented on Backend/ServiceLayer/Services/EmailService.cs:
```diff
@@ -0,0 +1,62 @@
+        private const string SmtpServer = "email-smtp.us-west-2.amazonaws.com";
+        private const int SmtpPort = 587;
+        private const string SmtpUsername = "AKIAIKGDZB4JHYJW4NHQ";
+        private const string SmtpPassword = "BDREGCLUMHfpqXy36czX4B9zQre6IbvCx4CausIn3pgQ";
```
These variables should come from the environment. By committing these, you've just made these public to the web. Anyone can now send mail from your AWS account. Please remove these from source and then *reset* these credentials in your AWS console.

In future, all configs should *not* be hardcoded. They *must* come from the environment.

This is **urgent**

julianpoy commented on Backend/ServiceLayer/Services/EmailService.cs:
```diff
@@ -0,0 +1,62 @@
+            return message;
+        }
+
+        public void sendEmail(MimeMessage messageToSend)
```
This seems flawed - The way it looks now, I have to call createEmailPlainBody and send it - then I have to call createEmailHTMLBody and send it - effectively sending two emails.

Something doesn't seem right here...

Good on you for recognizing that an email must contain a plaintext and an HTML part to them.

julianpoy commented on Backend/ServiceLayer/Services/ResetService.cs:
```diff
@@ -0,0 +1,229 @@
+        //Constant for the password reset email subject line
+        private const string resetPasswordSubjectString = "KFC SSO Password Reset";
+        //URL to append the reset ID to. Sent to the user in the email body.
+        private const string resetControllerURL = "kfcsso.com/api/reset/resetpassword/?id=";
```
This shouldn't be hardcoded. What if we change domains? Or another part of the stack changes?

This URL should be derived from the request headers coming in.

Perhaps leave off the domain part and build the rest like you are, using the heads to figure out that first part.

julianpoy commented on Backend/ServiceLayer/Services/ResetService.cs:
```diff
@@ -0,0 +1,229 @@
+            //Create the message that will be sent
+            MimeMessage emailToSend = es.createEmailPlainBody(userFullName, receiverEmail, resetPasswordSubjectString, resetPasswordBodyString);
+            //Send the email with the message
+            es.sendEmail(emailToSend);
```
As of now, looks like this is only sending a plaintext email - see my comment above in the email service.

julianpoy commented on Backend/ServiceLayer/Services/ResetService.cs:
```diff
@@ -0,0 +1,229 @@
+            string userFirstName = "";
+            string userlastName = "";
+            string userFullName = userFirstName + " " + userlastName;
+            string template = "Hi {0}, \r\n" +
```
Each of these templates should probably be located in a separate email template service or such. Whatever, this isn't that important.

julianpoy commented on Backend/ServiceLayer/Services/UserService.cs:
```diff
@@ -67,5 +67,10 @@ public User Login(DatabaseContext _db, string email, string password)
             return null;
         }
+
+        public bool ExistingUser(DatabaseContext _db, string email)
```
See my comment above regarding this method.

julianpoy commented on Backend/WebAPI/Controllers/ResetController.cs:
```diff
@@ -0,0 +1,123 @@
+
+        //After the user fills in the field with email, this action gets called
+        [HttpPost]
+        [ActionName("SendEmail")]
```
Call this action "send" for now.

julianpoy commented on Backend/WebAPI/Controllers/ResetController.cs:
```diff
@@ -0,0 +1,123 @@
+                    rs.AddResetIDToDB(email, resetID);
+                    string resetURL = rs.CreateResetURL(resetID);
+                    rs.SendResetEmailUserExists(email, resetID);
+                }
```
Remove this else condition. Everything within this else block is irrelevant.

julianpoy commented on Backend/WebAPI/Controllers/ResetController.cs:
```diff
@@ -0,0 +1,123 @@
+            if (umm.ExistingUser(email))
+            {
+                //Check to see if the account has a resetID already
+                if (!rs.CheckUserHasResetID(email)) //Don't have a resetID
```
Make the logic in this if block the default for every request - no need for the else.

julianpoy commented on Backend/WebAPI/Controllers/ResetController.cs:
```diff
@@ -0,0 +1,123 @@
+
+        //After the user clicks the link in the email, this action gets called
+        [HttpPost]
+        [ActionName("ResetPassword")]
```
Remove the name from this action. This can just be a POST request to the base URL.

julianpoy commented on Backend/WebAPI/Controllers/ResetController.cs:
```diff
@@ -0,0 +1,123 @@
+        //After the user clicks the link in the email, this action gets called
+        [HttpPost]
+        [ActionName("ResetPassword")]
+        public void ResetPassword(string resetID)
```
This accepts the "token".

julianpoy commented on Backend/WebAPI/Controllers/ResetController.cs:
```diff
@@ -0,0 +1,123 @@
+                }
+                else
+                {
+                    rs.DeleteResetIDFromDB(resetID);
```
Do not do this. Simply inform the user that the link has expired and redirect to the reset page. The token has expired in the DB, no need to delete it.

WinnMoo committed:
Updated code to requested changes

TODO:
Update PasswordResetManager to rollback DB changes
Finish ResetController

WinnMoo committed:
Updated PasswordResetManager to requested changes , added all unit tests, changes models

WinnMoo committed:
Added logic for creating a password, checking is resetToken is valid. Changed to environment variables for credentials.

julianpoy committed:
Revert 7652198a554317b2149939b645eb270e97b99f2f UserRepository
Whitespace

WinnMoo committed:
Reverted Configuration.cs to original state

WinnMoo committed:
Renamed PasswordResetManager --> PasswordManager, added more logic to PasswordManager, moved token generation to a service to allow other managers to utilize it, added another columns for PasswordReset model for resetting password

alfredovargasjr commented on Backend/UnitTesting/TestingUtils.cs:
```diff
@@ -90,7 +90,46 @@ public Session CreateSessionInDb(Session session)
+        }
+
+        /*
         public Service CreateServiceInDb(bool enabled)
```
Some of these unused utils and tests have been removed in the bug branch. Expect some merge conflicts

alfredovargasjr commented on Backend/WebAPI/Controllers/ResetController.cs:
```diff
@@ -0,0 +1,46 @@
+        //After the user fills in the field with email, this action gets called
+        [HttpPost]
+        [ActionName("Send")]
+        public void SendResetEmail(string email)
```
can you add the route header thing on top of each method so each route can be specified. 
[Route("api/reset")]

alfredovargasjr commented on Backend/WebAPI/Controllers/ResetController.cs:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        //After the user clicks the link in the email, this action gets called
+        [HttpPost]
```
same here for the route

WinnMoo committed:
Updated resetController with attribute routing

WinnMoo committed:
Added logic for sending reset email to user

julianpoy reviewed (dismissed):
At this point, I'm giving this the green light. You can merge when you see fit.

WinnMoo committed:
Merge branch 'feature/reset-password'

WinnMoo committed:
Merge branch 'master' of https://github.com/CECS-491A/SSO

WinnMoo committed:
Merge branch 'master' into feature/reset-password
